### PR TITLE
Loft Guidance Improvements + aamhybrid Guidance + QoL Changes

### PR DESCRIPTION
**Pull request by BillNyeTheIE**
- New field for loft guidance to decide terminal guidance, currently defaults to PN due to it performing best in testing
- Shaped charge warheads for missiles, shaped charge projectiles and rockets now show correct penetration

### CODE REVIEW

File: BDArmory/Guidances/MissileGuidance.cs [hunk 1]
```diff
@@ -156,7 +156,7 @@ public static Vector3 GetAirToGroundTarget(Vector3 targetPosition, Vector3 targe
             Vector3 targetAcceleration, Vessel missileVessel, float targetAlt, float maxAltitude,
             float rangeFactor, float vertVelComp, float velComp, float loftAngle, float termAngle,
             float termDist, ref int loftState, out float timeToImpact, out float targetDistance,
-            bool LoftUseAPN, float N, float minSpeed = 200)
+            MissileBase.GuidanceModes LoftGuidanceModeTerminal, float N, float minSpeed = 200)
         {
```
Comment (SuicidalInsanity) [inline BDArmory/Guidances/MissileGuidance.cs:160]: Just curious, but,how does this interact with standard terminalGuidance? What happens if both are setup/enabled on the same missile?

File: BDArmory/Weapons/Missiles/MultiMissileLauncher.cs [hunk 2]
```diff
@@ -333,6 +333,22 @@ void UpdateFields(MissileLauncher MLConfig, bool configurableSettings)
...
+                missileLauncher.LoftRangeFac = MLConfig.LoftRangeFac;
+                missileLauncher.LoftVelComp = MLConfig.LoftVelComp;
+                missileLauncher.LoftVertVelComp = MLConfig.LoftVertVelComp;
+                //missileLauncher.LoftAltComp = LoftAltComp;
+                missileLauncher.LoftTermRange = MLConfig.LoftTermRange;
+                missileLauncher.LoftGuidanceModeTerminal = MLConfig.LoftGuidanceModeTerminal;
+                missileLauncher.pronavGain = MLConfig.pronavGain;
+                missileLauncher.loftState = 0;
+                missileLauncher.TimeToImpact = float.PositiveInfinity;
+                missileLauncher.initMaxAoA = MLConfig.maxAoA;
```
Comment (SuicidalInsanity) [inline BDArmory/Weapons/Missiles/MultiMissileLauncher.cs:351]: if any of these settings are settings that are baked into the .cfg and/or internal values/not adjustable by the player in the SPH, they should go in the top section outside the if (configurableSettings) block to ensure they get properly passed along to the missile when updating fields.

File: BDArmory/Weapons/ModuleWeapon.cs [hunk 3]
```diff
@@ -5681,7 +5681,10 @@ public override string GetInfo()
                             output.AppendLine($"Cannister Round");
                             output.AppendLine($" - Submunition count: {binfo.subProjectileCount}");
                         }
-                        output.AppendLine($"Estimated Penetration: {ProjectileUtils.CalculatePenetration(binfo.caliber, binfo.bulletVelocity, binfo.bulletMass, binfo.apBulletMod, 940, 8.45001135e-07f, 0.656060636f, 1.20190930f, 1.77791929f):F2} mm");
+                        if (binfo.explosive == "shaped")
+                            output.AppendLine($"Estimated Penetration: {ProjectileUtils.CalculatePenetration(binfo.caliber > 0 ? binfo.caliber * 0.05f : 6f, 5000f, binfo.tntMass * 0.0555f, binfo.apBulletMod, 940, 0.00000094776185184f, 0.6560606203f, 1.201909309f, 1.777919321f):F2} mm");
+                        else
```
Comment (SuicidalInsanity) [inline BDArmory/Weapons/ModuleWeapon.cs:5686]: Might work better to have the shaped charge penetration be a separate line (" - Shaped Charge Penetration"?), on the off chance that someone makes custom bullets with a delay/penetrating fuzed shaped charge. Incredibly niche, and for most usage (e.g. barring naval shells(for which SC warheads would be of dubious usage anyway)) probably not going to add more than maybe 100mm penetration, so not a big deal either way, but something to consider.

File: BDArmory/Weapons/Missiles/MultiMissileLauncher.cs (hunk 2 above)
Comment (BillNyeTheIE) [reply BDArmory/Weapons/Missiles/MultiMissileLauncher.cs:351]: Right, good call.

File: BDArmory/Guidances/MissileGuidance.cs (hunk 1 above)
Comment (BillNyeTheIE) [reply BDArmory/Guidances/MissileGuidance.cs:160]: Shouldn't affect anything, terminalGuidance should only change the method by which the missile acquires the target, IE radar or IR, this is changing the guidance law when the missile goes terminal. The variable names for missile parameters could really use a re-organization, but given issues with compat I don't think we can do that.

File: BDArmory/Weapons/ModuleWeapon.cs (hunk 3 above)
Comment (BillNyeTheIE) [reply BDArmory/Weapons/ModuleWeapon.cs:5686]: I'll make that change, good call.

File: BDArmory/Guidances/MissileGuidance.cs (hunk 1 above)
Comment (BillNyeTheIE) [reply BDArmory/Guidances/MissileGuidance.cs:160]: On second thought, I'll be changing this variable name and LoftTermRange as well in order to introduce a new hybrid lead/PN guidance law.

File: BDArmory/Guidances/MissileGuidance.cs [hunk 4]
```diff
@@ -358,6 +359,40 @@ public static Vector3 GetAirToGroundTarget(Vector3 targetPosition, Vector3 targe
...
+            MissileBase.GuidanceModes homingModeTerminal, float N, float minSpeed = 200)
+        {
+            Vector3 velDirection = missileVessel.srf_vel_direction; //missileVessel.Velocity().normalized;
+
+            float targetDistance = Vector3.Distance(targetPosition, missileVessel.transform.position);
+
+            float currSpeed = Mathf.Max((float)missileVessel.srfSpeed, minSpeed);
+            Vector3 currVel = currSpeed * velDirection;
+
+            float leadTime = (float)(1 / ((targetVelocity - currVel).magnitude / targetDistance));
```
Comment (BrettRyland) [inline BDArmory/Guidances/MissileGuidance.cs:373]: This simplifies to
```csharp
float leadTime = targetDistance / (targetVelocity - currVel).magnitude;
```
Are there any potential issues if this returns `float.PositiveInfinity` for `timeToImpact`?

Comment (BillNyeTheIE) [reply BDArmory/Guidances/MissileGuidance.cs:373]: Not exactly sure why it's written like this, but that's how it's set up in multiple cases throughout the codebase. Should not cause major issues except a re-evaluation of whether or not it should loft. That won't cause major issues. The only time I would expect (targetVelocity - currVel).magnitude = 0 would be if the missile is just coming off the rails or if the missile has already spent all of its energy. In both cases a re-evaluation wouldn't affect anything.

**Pull request merged**